Optional Function returns break everything

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.7 63957f2.3307340
OS: linux, Ubuntu 24.04.1 LTS

getwd: ~/vlang/option_bug_maybe
vexe: ~/vlang/v/v
vexe mtime: 2024-09-21 23:21:32

vroot: OK, value: ~/vlang/v
VMODULES: OK, value: ~/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.43.0
Git vroot status: weekly.2024.37-36-g33073407
.git/config present: true

CC version: cc (Ubuntu 13.2.0-23ubuntu4) 13.2.0
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9

```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg src/main.v`
```v
module main

type DataFn = fn (name string) string

fn which_lang(name string) string {
	return name
}

fn find_func(name string) ?DataFn {
	return match name {
		'vlang' { which_lang }
		else { none }
	}
}

fn main() {
	req := find_func('vlang')
	fun := req('options')
	println(fun)
}

```

**What did you expect to see?**

a print of 'options'

**What did you see instead?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/main.01J8BE9J6F2YCCRKS27WSRJ220.tmp.c:13346: error: lvalue expected
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1197,6 +1197,18 @@ fn (mut c Checker) check_expr_option_or_result_call(expr ast.Expr, ret_type ast.
 					expr_ret_type = unaliased_ret_type
 				}
 			}
+			// var with option function
+			if expr.is_fn_var && expr.fn_var_type.has_option_or_result()
+				&& expr.or_block.kind == .absent {
+				ret_sym := c.table.sym(expr.fn_var_type)
+				if expr.fn_var_type.has_flag(.option) {
+					c.error('type `?${ret_sym.name}` is an Option, it must be unwrapped first',
+						expr.pos)
+				} else {
+					c.error('type `?${ret_sym.name}` is an Result, it must be unwrapped first',
+						expr.pos)
+				}
+			}
 			if expr_ret_type.has_option_or_result() {
 				if expr.or_block.kind == .absent {
 					ret_sym := c.table.sym(expr_ret_type)

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1090,7 +1090,7 @@ fn (mut c Checker) fn_call(mut node ast.CallExpr, mut continue_check &bool) ast.
 					} else {
 						if obj.typ == 0 {
 							if mut obj.expr is ast.IfGuardExpr {
-								typ = c.expr(mut obj.expr.expr)
+								typ = c.expr(mut obj.expr.expr).clear_option_and_result()
 							} else {
 								typ = c.expr(mut obj.expr)
 							}

--- a/vlib/v/checker/tests/option_fn_var_err.out
+++ b/vlib/v/checker/tests/option_fn_var_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/option_fn_var_err.vv:18:9: error: type `?DataFn` is an Option, it must be unwrapped first
+   16 | fn main() {
+   17 |     req := find_func('vlang')
+   18 |     fun := req('options')
+      |            ~~~~~~~~~~~~~~
+   19 |     println(fun)
+   20 | }

--- a/vlib/v/checker/tests/option_fn_var_err.vv
+++ b/vlib/v/checker/tests/option_fn_var_err.vv
@@ -1,0 +1,20 @@
+module main
+
+type DataFn = fn (name string) string
+
+fn which_lang(name string) string {
+	return name
+}
+
+fn find_func(name string) ?DataFn {
+	return match name {
+		'vlang' { which_lang }
+		else { none }
+	}
+}
+
+fn main() {
+	req := find_func('vlang')
+	fun := req('options')
+	println(fun)
+}

--- a/vlib/v/gen/c/match.v
+++ b/vlib/v/gen/c/match.v
@@ -92,10 +92,10 @@ fn (mut g Gen) match_expr(node ast.MatchExpr) {
 		cur_line = g.go_before_last_stmt().trim_left(' \t')
 		tmp_var = g.new_tmp_var()
 		mut func_decl := ''
-		if g.table.final_sym(node.return_type).kind == .function {
-			func_sym := g.table.final_sym(node.return_type)
-			if func_sym.info is ast.FnType {
-				def := g.fn_var_signature(func_sym.info.func.return_type, func_sym.info.func.params.map(it.typ),
+		ret_final_sym := g.table.final_sym(node.return_type)
+		if !node.return_type.has_option_or_result() && ret_final_sym.kind == .function {
+			if ret_final_sym.info is ast.FnType {
+				def := g.fn_var_signature(ret_final_sym.info.func.return_type, ret_final_sym.info.func.params.map(it.typ),
 					tmp_var)
 				func_decl = '${def} = &${g.typ(node.return_type)};'
 			}

--- a/vlib/v/tests/options/option_fn_var_test.v
+++ b/vlib/v/tests/options/option_fn_var_test.v
@@ -1,0 +1,21 @@
+module main
+
+type DataFn = fn (name string) string
+
+fn which_lang(name string) string {
+	return name
+}
+
+fn find_func(name string) ?DataFn {
+	return match name {
+		'vlang' { which_lang }
+		else { none }
+	}
+}
+
+fn test_main() {
+	req := find_func('vlang')?
+	fun := req('options')
+	println(fun)
+	assert fun == 'options'
+}

--- a/vlib/v/tests/options/option_fn_var_test.v
+++ b/vlib/v/tests/options/option_fn_var_test.v
@@ -19,3 +19,9 @@ fn test_main() {
 	println(fun)
 	assert fun == 'options'
 }
+
+fn test_ifguard() {
+	if req2 := find_func('vlang') {
+		assert req2('options') == 'options'
+	}
+}

